lisa/src/main/java/it/unive/lisa/analysis/heap/PointBasedHeap.java

```diff
@@ -0,0 +1,247 @@
+			return new PointBasedHeap(Collections.singleton(id), allocationSites);
+		}
+
+		if (expression instanceof HeapReference)
+			// TODO: not sure about this (need to check)
+			return this;
+
+		if (expression instanceof UnaryExpression) {
+			UnaryExpression unary = (UnaryExpression) expression;
+			PointBasedHeap sem = smallStepSemantics(unary.getExpression(), pp);
+			PointBasedHeap result = bottom();
+			for (ValueExpression expr : sem.getRewrittenExpressions())
+				result = result.lub(mk(sem, new UnaryExpression(expression.getTypes(), expr, unary.getOperator())));
+			return result;
+		}
+
+		if (expression instanceof BinaryExpression) {
+			BinaryExpression binary = (BinaryExpression) expression;
+			PointBasedHeap sem1 = smallStepSemantics(binary.getLeft(), pp);
+			PointBasedHeap sem2 = sem1.smallStepSemantics(binary.getRight(), pp);
+			PointBasedHeap result = bottom();
+			for (ValueExpression expr1 : sem1.getRewrittenExpressions())
+				for (ValueExpression expr2 : sem2.getRewrittenExpressions())
+					result = result.lub(
+							mk(sem2, new BinaryExpression(expression.getTypes(), expr1, expr2, binary.getOperator())));
+			return result;
+		}
+
+		if (expression instanceof ValueIdentifier)
+			if (allocationSites.containsKey(expression))
+				return new PointBasedHeap(allocationSites.get(expression), allocationSites);
+
+		if (expression instanceof ValueExpression)
+			return mk(this, (ValueExpression) expression);
+
+		return top();
+	}
+
+	@Override
+	public PointBasedHeap assign(Identifier id, SymbolicExpression expression, ProgramPoint pp)
+			throws SemanticException {
+
+		if (expression instanceof HeapIdentifier) {
+			HashMap<Identifier, Collection<ValueExpression>> sites = new HashMap<>(allocationSites);
+			HashSet<ValueExpression> v = new HashSet<>();
+			v.add((ValueExpression) expression);
+			sites.put(id, v);
+			return new PointBasedHeap(Collections.singleton((ValueExpression) expression), sites);
+		}
+
+		return smallStepSemantics(expression, pp);
+	}
+
+	@Override
+	public PointBasedHeap assume(SymbolicExpression expression, ProgramPoint pp) throws SemanticException {
+		// we just rewrite the expression if needed
+		return smallStepSemantics(expression, pp);
+	}
+
+	@Override
+	public PointBasedHeap forgetIdentifier(Identifier id) throws SemanticException {
+		Map<Identifier, Collection<
+				ValueExpression>> sites = new HashMap<Identifier, Collection<ValueExpression>>(allocationSites);
+		sites.remove(id);
+		return new PointBasedHeap(rewritten, sites);
+	}
+
+	@Override
+	public Satisfiability satisfies(SymbolicExpression expression, ProgramPoint pp) throws SemanticException {
+		// we leave the decision to the value domain
+		return Satisfiability.UNKNOWN;
+	}
+
+	@Override
+	public String representation() {
+		HashSet<ValueExpression> res = new HashSet<ValueExpression>();
+		for (Collection<ValueExpression> s : allocationSites.values())
```

Comment:
Since this will end up in dot files, comparison between graph might break if two instances of this domain contains the same data but the representation joins it in different order (ex `"a, b".equals("b, a") == false`) but for the logic of this domain order does not matter. We have to enforce ordering on the element being dumped. Something like changing res to a collection (**not** a specific instance like HashSet) of strings and then using a TreeSet, and instead of adding the expressions we directly add their `toString()` value.